File: packages/oracle-host-node/src/live/h3-helper/protocol.ts
/**
 * The framed stdio protocol's TypeScript twin — frame type ids and
 * control payload shapes. The contract is
 * the request-engine H3-protocol design; the Rust twin is
 * `native/h3-helper/src/protocol.rs`. Any wire-shape change bumps
 * {@link H3_PROTOCOL_VERSION} on both sides — no negotiation, both
 * sides ship from one repo.
 */

export const H3_PROTOCOL_VERSION = 3;

/** The `'3'` pipeline's entire legal cipher vocabulary — QUIC is TLS
 *  1.3-only and the helper's provider carries exactly these three
 *  suites, so the pre-wire gate admits these IANA names and nothing
 *  else (no OpenSSL-format lists, no other TLS 1.3 names). */
export const H3_TLS13_CIPHER_SUITES = [
  'TLS_AES_128_GCM_SHA256',
  'TLS_AES_256_GCM_SHA384',
  'TLS_CHACHA20_POLY1305_SHA256',
] as const;

export const H3_FRAME = {
  HELLO: 0x01,
  REQUEST: 0x10,
  REQUEST_BODY: 0x11,
  REQUEST_END: 0x12,
  CANCEL: 0x1f,
  RESPONSE_HEAD: 0x20,
  RESPONSE_BODY: 0x21,
  RESPONSE_TRAILERS: 0x22,
  RESPONSE_END: 0x23,
  ERROR: 0x2e,
} as const;

export type H3HeaderPair = [name: string, value: string];

export interface H3Hello {
  protocol: number;
  helper: string;
}

export interface H3ClientCert {
  certPem: string;
  /** PKCS#8 PEM, already decrypted node-side — a passphrase never
   *  crosses the protocol (rustls cannot decrypt encrypted PEM). */
  keyPem: string;
}

export interface H3RequestHead {
  /** https:// only — the transport guards cleartext pre-wire. */
  url: string;
  method: string;
  /** User-set Host header, already folded out of `headers` — overrides
   *  the URL's host as `:authority` (the h2 pipeline's contract). */
  authority?: string;
  /** Wire order, repeats allowed; connection-specific fields already
   *  stripped node-side (the h2 hygiene rule). */
  headers: H3HeaderPair[];
  /** Announced body length. 0 = the head is complete by itself;
   *  otherwise REQUEST_BODY frames totalling exactly this, then
   *  REQUEST_END. */
  bodyBytes: number;
  /** `sslVerification: false` — the helper accepts any certificate. */
  insecure?: true;
  clientCert?: H3ClientCert;
  /** resolveToAddress pin — the helper dials this IP; SNI and
   *  certificate verification keep the URL's host. */
  connectAddress?: string;
  /** TLS 1.3 IANA suite names restricting the handshake's offer —
   *  members of {@link H3_TLS13_CIPHER_SUITES} only, gated pre-wire. */
  cipherSuites?: string[];
  /** QUIC max-idle ceiling; helper default 30 000. */
  idleTimeoutMs?: number;
  /** `captureNetwork` (v3) — the helper dials a fresh instrumented
   *  connection (never a pooled one) and reports its socket facts +
   *  dial timings on RESPONSE_HEAD. */
  captureNetwork?: true;
}

/** Endpoints of the instrumented dial's socket (v3) — the UDP socket
 *  the QUIC connection rides. */
export interface H3SocketFacts {
  localAddress: string;
  localPort: number;
  remoteAddress: string;
  remotePort: number;
}

/** Phase durations of the instrumented dial (v3). QUIC merges
 *  transport establishment and TLS into ONE handshake, so there is no
 *  TCP-connect leg — `handshakeMs` is the whole of it. */
export interface H3DialTimings {
  /** Absent when `connectAddress` pinned the dial — nothing resolved. */
  dnsMs?: number;
  handshakeMs: number;
}

export interface H3ResponseHead {
  status: number;
  headers: H3HeaderPair[];
  /** v3, present only on a `captureNetwork` head's instrumented dial. */
  socket?: H3SocketFacts;
  /** v3, present only on a `captureNetwork` head's instrumented dial. */
  timings?: H3DialTimings;
}

/** Helper-side codes are a closed set per protocol version; the node
 *  classifier maps them and falls back generically on any it doesn't
 *  know (forward-compatibility across a version bump in dev). */
export interface H3ErrorFrame {
  code: string;
  message: string;
}

File: native/h3-helper/src/protocol.rs
//! The framed stdio protocol's Rust twin — frame type ids and control
//! payload shapes. The contract is `docs/REQUEST_ENGINE_H3_PROTOCOL.md`;
//! the TypeScript twin is `oracle-host-node/src/live/h3-helper/protocol.ts`.
//! Any wire-shape change bumps `PROTOCOL_VERSION` on both sides.

use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 3;

pub mod frame_type {
    pub const HELLO: u8 = 0x01;
    pub const REQUEST: u8 = 0x10;
    pub const REQUEST_BODY: u8 = 0x11;
    pub const REQUEST_END: u8 = 0x12;
    pub const CANCEL: u8 = 0x1f;
    pub const RESPONSE_HEAD: u8 = 0x20;
    pub const RESPONSE_BODY: u8 = 0x21;
    pub const RESPONSE_TRAILERS: u8 = 0x22;
    pub const RESPONSE_END: u8 = 0x23;
    pub const ERROR: u8 = 0x2e;
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Hello {
    pub protocol: u32,
    pub helper: &'static str,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientCert {
    pub cert_pem: String,
    /// PKCS#8 PEM, already decrypted node-side — a passphrase never
    /// crosses the protocol.
    pub key_pem: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RequestHead {
    pub url: String,
    pub method: String,
    /// User-set Host header, already folded out of `headers` node-side —
    /// overrides the URL's host as `:authority`.
    #[serde(default)]
    pub authority: Option<String>,
    pub headers: Vec<(String, String)>,
    pub body_bytes: u64,
    #[serde(default)]
    pub insecure: bool,
    #[serde(default)]
    pub client_cert: Option<ClientCert>,
    /// resolveToAddress pin — dial this IP; SNI and certificate
    /// verification keep the URL's host.
    #[serde(default)]
    pub connect_address: Option<String>,
    /// TLS 1.3 IANA suite names restricting the handshake's offer —
    /// the node side admits only names this helper's provider carries.
    #[serde(default)]
    pub cipher_suites: Option<Vec<String>>,
    #[serde(default)]
    pub idle_timeout_ms: Option<u64>,
    /// captureNetwork (v3): dial a fresh instrumented connection —
    /// never a pooled one — and report its socket facts + dial timings
    /// on RESPONSE_HEAD.
    #[serde(default)]
    pub capture_network: bool,
}

/// Endpoints of the instrumented dial's socket (v3) — the UDP socket
/// the QUIC connection rides.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SocketFacts {
    pub local_address: String,
    pub local_port: u16,
    pub remote_address: String,
    pub remote_port: u16,
}

/// Phase durations of the instrumented dial (v3). QUIC merges
/// transport establishment and TLS into ONE handshake, so there is no
/// TCP-connect leg — `handshake_ms` is the whole of it.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DialTimings {
    /// Absent when `connectAddress` pinned the dial — nothing resolved.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dns_ms: Option<f64>,
    pub handshake_ms: f64,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResponseHead {
    pub status: u16,
    pub headers: Vec<(String, String)>,
    /// v3, present only on a `captureNetwork` head's instrumented dial.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub socket: Option<SocketFacts>,
    /// v3, present only on a `captureNetwork` head's instrumented dial.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timings: Option<DialTimings>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ErrorFrame {
    pub code: &'static str,
    pub message: String,
}
